interop-testing/src/main/java/io/grpc/testing/integration/XdsTestClient.java

```diff
@@ -0,0 +1,336 @@
+      System.err.println(
+          "Usage: [ARGS...]"
+              + "\n"
+              + "\n  --num_channels=INT     Default: "
+              + c.numChannels
+              + "\n  --print_response=BOOL  Write RPC response to stdout. Default: "
+              + c.printResponse
+              + "\n  --qps=INT              Qps per channel. Default: "
+              + c.qps
+              + "\n  --rpc_timeout_sec=INT  Per RPC timeout seconds. Default: "
+              + c.rpcTimeoutSec
+              + "\n  --server=host:port     Address of server. Default: "
+              + c.server
+              + "\n  --stats_port=INT       Port to expose peer distribution stats service. "
+              + "Default: "
+              + c.statsPort);
+      System.exit(1);
+    }
+  }
+
+  private void run() {
+    statsServer = NettyServerBuilder.forPort(statsPort).addService(new XdsStatsImpl()).build();
+    boolean success = false;
+    try {
+      statsServer.start();
+      List<ManagedChannel> chans = new ArrayList<>();
+      for (int i = 0; i < numChannels; i++) {
+        chans.add(NettyChannelBuilder.forTarget(server).usePlaintext().build());
+      }
+      ListeningScheduledExecutorService exec =
+          MoreExecutors.listeningDecorator(Executors.newSingleThreadScheduledExecutor());
+      try {
+        runQps(chans, exec);
+        success = true;
+      } finally {
+        exec.shutdownNow();
+        for (ManagedChannel chan : chans) {
+          chan.shutdownNow();
+        }
+      }
+    } catch (Throwable t) {
+      logger.log(Level.SEVERE, "Error running client", t);
+    }
+
+    statsServer.shutdownNow();
+    System.exit(success ? 0 : 1);
+  }
+
+  private void stop() throws Exception {
+    statsServer.shutdownNow();
+    if (!statsServer.awaitTermination(5, TimeUnit.SECONDS)) {
+      System.err.println("Timed out waiting for server shutdown");
+    }
+  }
+
+  private final Set<XdsStatsWatcher> watchers = new HashSet<>();
+  private final Object lock = new Object();
+  private long currentRequestId;
+
+  private void runQps(final List<ManagedChannel> chans, ListeningScheduledExecutorService exec)
+      throws InterruptedException, ExecutionException, TimeoutException {
+    final SettableFuture<Void> failure = SettableFuture.create();
+    final class PeriodicRpc implements Runnable {
+      final AtomicLong messageIds = new AtomicLong();
+
+      @Override
+      public void run() {
+        final long requestId;
+        final Set<XdsStatsWatcher> savedWatchers = new HashSet<>();
+        synchronized (lock) {
+          currentRequestId += 1;
+          requestId = currentRequestId;
+          savedWatchers.addAll(watchers);
+        }
+
+        SimpleRequest request = SimpleRequest.newBuilder().setFillServerId(true).build();
+        ManagedChannel channel = chans.get((int) (requestId % chans.size()));
+        final ClientCall<SimpleRequest, SimpleResponse> call =
+            channel.newCall(
```

Comment:
Why not just call the rpc method with stub?

Reply:
To be able to invoke `call.getAttributes().get(Grpc.TRANSPORT_ATTR_REMOTE_ADDR)`. If there's a simpler way to get access to the call attributes I'd be happy to change this.

Reply:
An alternative is using interceptor, it may not be much simpler though.

```java
AtomicReference<ClientCall> clientCallRef = new AtomicReference<>();
stub = stub.withDeadlineAfter().withInterceptors(
  new ClientInterceptor() {
      @Override
      public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
          MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel next) {
        ClientCall<ReqT, RespT> call = next.newCall(method,callOptions);
        clientCallRef.set(call);
        return call;
  });

try {
  SimpleResponse response = stub.unaryCall(request);
  serverId = response.getServerId();
  if (printResponse) {
    System.out.println(
       "Greeting: Hello world, this is "
          + response.getHostname()
          + ", from "
          + clientCallRef.get().getAttributes().get(Grpc.TRANSPORT_ATTR_REMOTE_ADDR));
  }
} catch(RuntimeException e) {
} finally {
  for (XdsStatsWatcher watcher : savedWatchers) {
    watcher.rpcCompleted(requestId, serverId);
  }
}
```

Reply:
Yeah I'm not sure that's simpler either :) I think it's "equally" valid to directly interact with the ClientCall API as opposed to using an interceptor (there are at least examples of how do so in the ClientCall javadoc)?